charts/core/src/Echart/useEchart.ts

```diff
@@ -0,0 +1,386 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import debounce from 'lodash.debounce';
+
+import { ChartOptions } from '../Chart';
+import { chartSeriesColors } from '../Chart/chartSeriesColors';
+import * as updateUtils from '../Chart/hooks/updateUtils';
+
+import {
+  EChartHookProps,
+  EChartsInstance,
+  ZoomSelectionEvent,
+} from './Echart.types';
+
+// Singleton promise for initialization to prevent duplication
+let initPromise: Promise<void> | null = null;
+let echartsCore: any;
+let echartsCharts: any;
+let echartsRenders: any;
+let echartsComponents: any;
+
+async function initializeEcharts() {
+  // If already initialized, return immediately
+  if (echartsCore) {
+    return;
+  }
+
+  // If initialization is in progress, wait for it
+  if (initPromise) {
+    return initPromise;
+  }
+
+  // Start initialization and store the promise
+  initPromise = (async () => {
+    try {
+      const [
+        echartsCoreResolved,
+        echartsChartsResolved,
+        echartsRendersResolved,
+        echartsComponentsResolved,
+      ] = await Promise.all([
+        import('echarts/core'),
+        import('echarts/charts'),
+        import('echarts/renderers'),
+        import('echarts/components'),
+      ]);
+
+      echartsCore = echartsCoreResolved;
+      echartsCharts = echartsChartsResolved;
+      echartsRenders = echartsRendersResolved;
+      echartsComponents = echartsComponentsResolved;
+
+      echartsCore.use([
+        echartsCharts.LineChart,
+        echartsRenders.CanvasRenderer,
+        echartsComponents.TitleComponent,
+        echartsComponents.TooltipComponent,
+        echartsComponents.GridComponent,
+        echartsComponents.LegendComponent,
+        echartsComponents.ToolboxComponent,
+        echartsComponents.DataZoomComponent,
+        echartsComponents.DataZoomInsideComponent,
+      ]);
+    } catch (error) {
+      // Ensure we clear the promise if initialization fails
+      initPromise = null;
+      throw error;
+    }
+  })();
+
+  return initPromise;
+}
+
+/**
+ * Wrapper around the ECharts library. Instantiates an ECharts instance.
+ * Provides helper methods to hide ECharts specific logic and give a cleaner API
+ * for interacting with a chart.
+ */
+export function useEchart({
+  container,
+  initialOptions,
+  theme,
+}: EChartHookProps): EChartsInstance {
+  const [echartsInstance, setEchartsInstance] = useState<any>(null); // has to be any since no types exist until import
+  const [error, setError] = useState<EChartsInstance['error']>(null);
+  const [ready, setReady] = useState<EChartsInstance['ready']>(false);
+  const [options, setOptions] = useState<EChartsInstance['options']>(
```

Comment:
remind me, why do we need to keep track of options in state if we can access them via echartsInstance.options?

Reply:
If we want it more easily accessible from this hook's return object, could we just have
```ts
const options = useMemo(() => echartsInstance.options, [...]) // or whatever the correct property is?
```

Reply:
I _think_ at one point early on I was tying into changes in order to trigger updates. Since this is no longer the case, I think this is a really good point. Will look into removing this.

Reply:
Okay I dug into this more and remember why we were holding a copy of the options. This is what made it possible to batch updates. Since we debounce calls to `echartsInstance.setOption()` via `setEchartOptions`, when additional updates occur in short order, they don't have an accurate copy of the current options since `echartsInstance.getOption()` is not necessarily up to date yet.

I know there was some initial disagreement regarding whether this batching was necessary. I tried to run some tests rendering a chart with 100 lines. With batching, the chart appeared after about 4000ms. Without batching, it didn't appear until after the profiler stopped recording which seemed to be around 6000ms. I'm not 100% sure what caused this but I am under the impression that each additional update is supposed to cause a repaint of the canvas, so it would be 1 repaint vs 100+ repaints. Though, I have no idea if that's what caused this, it could be something totally unrelated, at least withe current implementation it caused an inherent slowdown.

Happy to discuss further of course if there are other ideas, but just going to leave as is for now.